libs/server/Objects/Set/SetObjectImpl.cs

```diff
@@ -256,5 +258,115 @@ private void SetPop(byte* input, int length, ref SpanByteAndMemory output)
+
+            int countDone = 0;
+            bool isMemory = false;
+            MemoryHandle ptrHandle = default;
+            byte* ptr = output.SpanByte.ToPointer();
+
+            var curr = ptr;
+            var end = curr + output.Length;
+
+            ObjectOutputHeader _output = default;
+
+            try
+            {
+                int[] indexes = default;
+
+                if (count > 0)
+                {
+                    // Return an array of distinct elements
+                    var countParameter = count > set.Count ? set.Count : count;
+                    List<byte[]> randomElements = new List<byte[]>();
+
+                    // The order of fields in the reply is not truly random
+                    indexes = new HashSet<int>(Enumerable.Range(0, set.Count).OrderBy(x => Guid.NewGuid()).Take(countParameter)).ToArray();
+
+                    // Write the size of the array reply
+                    while (!RespWriteUtils.WriteArrayLength(countParameter, ref curr, end))
+                        ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+
+                    foreach (var index in indexes)
+                    {
+                        var element = set.ElementAt(index);
+                        while (!RespWriteUtils.WriteBulkString(element, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                        countDone++;
+                    }
+                    countDone += count - countParameter;
+                }
+                else if (count == int.MinValue) // no count parameter is present
+                {
+                    // Return a single random element from the set
+                    if (set.Count > 0)
+                    {
+                        int index = RandomNumberGenerator.GetInt32(0, set.Count);
+                        var item = set.ElementAt(index);
+                        while (!RespWriteUtils.WriteBulkString(item, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                    }
+                    else
+                    {
+                        // If set is empty, return nil
+                        while (!RespWriteUtils.WriteDirect(CmdStrings.RESP_ERRNOTFOUND, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                    }
+                    countDone++;
+                }
+                else // count < 0
+                {
+                    // Return an array with potentially duplicate elements
+                    int countParameter = Math.Abs(count);
+
+                    Random random = new Random();
+                    List<int> randomIndexes = new List<int>();
```

Comment:
Instead of adding to a list and then converting to an array, you could add the values directly to an array, since you already know the array size in advance